pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
 
 func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
 	if pl.enableReadWriteOncePod {
-		return pl.isReadWriteOncePodAccessModeConflict(pod)
+		return pl.isReadWriteOncePodAccessModeConflict(cycleState, pod)
 	}
+	p := &preFilterState{}
+	cycleState.Write(preFilterStateKey, p)
 	return framework.NewStatus(framework.Success)
 }
 
 // isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
 // This access mode restricts volume access to a single pod on a single node. Since only a single pod can
 // use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
-// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
-func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
 	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
 	if err != nil {
 		return framework.NewStatus(framework.Error, "error while getting node info")
 	}
+	pvcKeys, status := pl.keysForPVCWithRWOP(pod)
+	if status != nil {
+		return status
+	}
+	var nodesMatchPVCWithRWOP []*v1.Node
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeInfos[i]
+		pvcKeysNumber := uint32(len(pvcKeys))
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				atomic.AddUint32(&pvcKeysNumber, ^uint32(0))
+				if pvcKeysNumber == 0 {
+					nodesMatchPVCWithRWOP = append(nodesMatchPVCWithRWOP, nodeInfo.Node())
+					cancel()
+				}
+			}
+		}
+	}
+	pl.parallelizer.Until(ctx, len(nodeInfos), processNode)
+	// update cyclestate
```

Comment:
nit: Can you provide more detail here on why we need to store this information?

Reply:
> nit: Can you provide more detail here on why we need to store this information or how it will be used later?

https://github.com/kubernetes/kubernetes/blob/acfdb93f752a3635f41aa177944b81b8512f3cb2/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L345-L355

I have explained the fields here.